rand_pcg/src/pcg128cm.rs

```diff
@@ -0,0 +1,183 @@
+    ///
+    /// PCG specifies the following default values for both parameters:
+    ///
+    /// - `state = 0xcafef00dd15ea5e5`
+    /// - `stream = 0xa02bdbf7bb3c0a7ac28fa16a64abf96`
+    ///
+    /// [upgrading-pcg64]: https://numpy.org/doc/stable/reference/random/upgrading-pcg64.html
+    pub fn new(state: u128, stream: u128) -> Self {
+        // The increment must be odd, hence we discard one bit:
+        let increment = (stream << 1) | 1;
+        Self::from_state_incr(state, increment)
+    }
+
+    #[inline]
+    fn from_state_incr(state: u128, increment: u128) -> Self {
+        let mut pcg = Self { state, increment };
+        // Move away from initial value:
+        pcg.state = pcg.state.wrapping_add(pcg.increment);
+        pcg.step();
+        pcg
+    }
+
+    #[inline]
+    fn step(&mut self) {
+        // prepare the LCG for the next round
+        self.state = self
+            .state
+            .wrapping_mul(MULTIPLIER as u128)
+            .wrapping_add(self.increment);
+    }
+}
+
+// Custom Debug implementation that does not expose the internal state
+impl fmt::Debug for Lcg128CmDxsm64 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Lcg128CmDxsm64 {{}}")
+    }
+}
+
+/// We use a single 255-bit seed to initialise the state and select a stream.
+/// One `seed` bit (lowest bit of `seed[8]`) is ignored.
+impl SeedableRng for Lcg128CmDxsm64 {
+    type Seed = [u8; 32];
+
+    fn from_seed(seed: Self::Seed) -> Self {
+        let mut seed_u64 = [0u64; 4];
+        le::read_u64_into(&seed, &mut seed_u64);
+        let state = u128::from(seed_u64[0]) | (u128::from(seed_u64[1]) << 64);
+        let incr = u128::from(seed_u64[2]) | (u128::from(seed_u64[3]) << 64);
+
+        // The increment must be odd, hence we discard one bit:
+        Self::from_state_incr(state, incr | 1)
+    }
+}
+
+impl RngCore for Lcg128CmDxsm64 {
+    #[inline]
+    fn next_u32(&mut self) -> u32 {
+        self.next_u64() as u32
+    }
+
+    #[inline]
+    fn next_u64(&mut self) -> u64 {
+        let val = output_dxsm(self.state);
+        self.step();
+        val
+    }
+
+    #[inline]
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        impls::fill_bytes_via_next(self, dest)
+    }
+
+    #[inline]
+    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), Error> {
+        self.fill_bytes(dest);
+        Ok(())
+    }
+}
+
+#[inline(always)]
+fn output_dxsm(state: u128) -> u64 {
+    // This permutation is similar to xorshift multiply hash functions, except
+    // that one of the multipliers is the LCG multiplier (to avoid needing to
+    // have a second constant) and the other is based on the low-order bits.
+    // This latter aspect means that the scrambling applied to the high bits
+    // depends on the low bits, and makes it (to my eye) impractical to back
+    // out the permutation without having the low-order bits.
```

Comment:
Is that your comment or is it taken from the reference? I would like to avoid "impractical", because it may be misunderstood to suggest cryptographic security.

Reply:
> Is that your comment or is it taken from the reference?

It is taken from pcg-cpp's source code.

Personally, I would alright with dropping the comment entirely as pcg-cpp's source is available for reference in any case. If you want, I could also just remove "impractical" or explicitly mark the whole comment as a quote?

Reply:
I see. Maybe let's remove the comment then and put a link to the reference instead?